pkg/operators/wasm/perf.go

```diff
@@ -0,0 +1,242 @@
+// - stack[3] is the value size
+// - stack[4] is the max entries
+// Return value:
+// - Perf reader handle on success, 0 on error
+func (i *wasmOperatorInstance) newPerfReader(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfMapHandle := wapi.DecodeU32(stack[0])
+	size := int(wapi.DecodeU32(stack[1]))
+	isOverwritable := wapi.DecodeU32(stack[2]) > 0
+
+	perfMap, ok := getHandle[*ebpf.Map](i, perfMapHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	perfReader, err := perf.NewReaderWithOptions(perfMap, size, perf.ReaderOptions{Overwritable: isOverwritable})
+	if err != nil {
+		i.logger.Warnf("newPerfReader: creating perf reader: %v", err)
+		stack[0] = 0
+		return
+	}
+
+	stack[0] = wapi.EncodeU32(i.addHandle(perfReader))
+}
+
+// perfReaderPause() pauses the perf reader
+// Params:
+// - stack[0]: Perf reader handle
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) perfReaderPause(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfReaderHandle := wapi.DecodeU32(stack[0])
+
+	perfReader, ok := getHandle[*perf.Reader](i, perfReaderHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	err := perfReader.Pause()
+	if err != nil {
+		i.logger.Warnf("perfReaderPause: pausing perf reader: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	stack[0] = 0
+}
+
+// perfReaderResume() resume the perf reader
+// Params:
+// - stack[0]: Perf reader handle
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) perfReaderResume(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfReaderHandle := wapi.DecodeU32(stack[0])
+
+	perfReader, ok := getHandle[*perf.Reader](i, perfReaderHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	err := perfReader.Resume()
+	if err != nil {
+		i.logger.Warnf("perfReaderResume: resuming perf reader: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	stack[0] = 0
+}
+
+// perfReaderSetDeadline() set the perf reader deadline
+// Params:
+// - stack[0]: Perf reader handle
+// - stack[1]: Time in nanoseconds
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) perfReaderSetDeadline(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfReaderHandle := wapi.DecodeU32(stack[0])
+	nsec := stack[1]
+
+	perfReader, ok := getHandle[*perf.Reader](i, perfReaderHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	perfReader.SetDeadline(time.Unix(0, int64(nsec)))
+
+	stack[0] = 0
+}
+
+// perfReaderRead read one record from the perf reader.
+// Params:
+// - stack[0]: Perf reader handle
+// - stack[1]: bufPtr address
+// Return value:
+// - 0 on success, 1 on error, 2 on deadline exceeded
+func (i *wasmOperatorInstance) perfReaderRead(ctx context.Context, m wapi.Module, stack []uint64) {
+	perfReaderHandle := wapi.DecodeU32(stack[0])
+	addrBufPtr := wapi.DecodeU32(stack[1])
+
+	perfReader, ok := getHandle[*perf.Reader](i, perfReaderHandle)
+	if !ok {
+		stack[0] = 1
+		return
+	}
+
+	record, err := perfReader.Read()
```

Comment:
I still have concerns about exposing blocking functions to the wasm module.

Even if traceloop uses `perfReaderRead` in a non-blocking way (the ring buffer is configured as backward and overwriteable. AFAIK, that makes it non-blocking), this new API enables blocking methods.

Instead of mirroring the cilium/ebpf/perf API into wasm, I would prefer to have a higher-level API without the blocking behaviour. In practice, gadgets/traceloop/go/program.go would not call Pause+SetDeadline+Read+Resume but a single call to the wasm host doing the 4 function calls in one step. In this way, the wasm module would not be able to block.

Previous discussion in https://github.com/inspektor-gadget/inspektor-gadget/pull/3714#discussion_r1856381535